### PR TITLE
tbb 2021.2.0

### PR DESCRIPTION
**Pull request by bayandin**
Created with `brew bump-formula-pr`.

I'm not sure if I need to do anything with the version. 
Before it was `2020_U3`, but now there is no reference to such version style. Also, the repo has been moved, so they could also have changed the versioning style.

### CODE REVIEW

**Comment by carlocab** [comment]
Patch no longer applies.

**Comment by bayandin** [comment]
> Patch no longer applies.

The patch changes `build/macos.inc`, but `build` directory doesn't exist anymore (it was deleted in a commit with a lot of other stuff labelled as [`Committing oneTBB 2021.1-beta08 source code`](https://github.com/oneapi-src/oneTBB/commit/05b1dd0789a02dfa77fa802b7172f52e02e2d788), so it's hard to say what was really changed).

Let's try just to delete the patch and see if it magically works on Apple Silicon.

**Comment by bayandin** [comment]
```
brew test --retry --verbose bowtie2
brew test --retry --verbose caffe
brew test --retry --verbose ceres-solver
brew test --retry --verbose dynare
brew test --retry --verbose embree
brew test --retry --verbose gmsh
brew test --retry --verbose open-image-denoise
brew test --retry --verbose opencascade
brew test --retry --verbose opencv
brew test --retry --verbose opencv@3
brew test --retry --verbose openvdb
brew test --retry --verbose ospray
brew test --retry --verbose osrm-backend
brew test --retry --verbose pagmo
brew test --retry --verbose parallelstl
brew test --retry --verbose root
brew test --retry --verbose siril
brew test --retry --verbose sysdig
```
Most of them could be fixed by a revision bum, they fails with error like: 
```
dyld: Symbol not found: __ZN3tbb4task13note_affinityEt
  Referenced from: /usr/local/opt/embree/lib/libembree3.3.dylib
  Expected in: /usr/local/opt/tbb/lib/libtbb.dylib
 in /usr/local/opt/embree/lib/libembree3.3.dylib
```

But some of them, probably, aren't compatible with `tbb 2021.1.1` yet, like `parallelstl`:
```
==> /usr/bin/clang++ -std=c++11 -L/usr/local/opt/tbb/lib -ltbb -I/usr/local/Cellar/parallelstl/20201111/stdlib -I/usr/local/Cellar/parallelstl/20201111/include test.cpp -o test
In file included from test.cpp:1:
In file included from /usr/local/Cellar/parallelstl/20201111/stdlib/pstl/execution:44:
In file included from /usr/local/Cellar/parallelstl/20201111/include/pstl/internal/glue_execution_defs.h:58:
In file included from /usr/local/Cellar/parallelstl/20201111/include/pstl/internal/algorithm_impl.h:32:
In file included from /usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend.h:28:
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:82:10: error: no member named 'task' in namespace 'tbb'
    tbb::task::self().group()->cancel_group_execution();
    ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:427:34: error: expected class name
class __merge_task : public tbb::task
                                 ^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:434:23: error: no type named 'task' in namespace 'tbb'
    /*override*/ tbb::task*
                 ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:625:10: error: no type named 'task' in namespace 'tbb'
    tbb::task*
    ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:664:10: error: no type named 'task' in namespace 'tbb'
    tbb::task*
    ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:711:10: error: no type named 'task' in namespace 'tbb'
    tbb::task*
    ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:564:14: error: no type named 'task' in namespace 'tbb'
        tbb::task* p = (_root ? nullptr : parent());
        ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:564:43: error: use of undeclared identifier 'parent'
        tbb::task* p = (_root ? nullptr : parent());
                                          ^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:746:43: error: no member named 'task' in namespace 'tbb'
        __merge_task* __right = new (tbb::task::allocate_additional_child_of(*parent()))
                                     ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:746:79: error: use of undeclared identifier 'parent'
        __merge_task* __right = new (tbb::task::allocate_additional_child_of(*parent()))
                                                                              ^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:752:14: error: no member named 'task' in namespace 'tbb'
        tbb::task::spawn(*__right);
        ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:753:14: error: no member named 'task' in namespace 'tbb'
        tbb::task::recycle_as_continuation();
        ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:764:6: error: no type named 'task' in namespace 'tbb'
tbb::task*
~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:805:40: error: expected class name
class __stable_sort_task : public tbb::task
                                       ^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:813:23: error: no type named 'task' in namespace 'tbb'
    /*override*/ tbb::task*
                 ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:835:6: error: no type named 'task' in namespace 'tbb'
tbb::task*
~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:853:14: error: no type named 'task' in namespace 'tbb'
        tbb::task* p = parent();
        ~~~~~^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:853:24: error: use of undeclared identifier 'parent'
        tbb::task* p = parent();
                       ^
/usr/local/Cellar/parallelstl/20201111/include/pstl/internal/parallel_backend_tbb.h:862:32: error: use of undeclared identifier 'allocate_continuation'
    _MergeTaskType* __m = new (allocate_continuation()) _MergeTaskType(
                               ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

Will take a closer look later

**Comment by NeighbourDYY** [comment]
I just randomly came across this discussion. It’s known that GCC would not be compatible with this version of tbb. See the Known Limitations in
https://software.intel.com/content/www/us/en/develop/articles/intel-oneapi-threading-building-blocks-release-notes.html

Hope it’s helpful.

**Comment by bayandin** [comment]
- `opencv` added support for `tbb` 2021 in https://github.com/opencv/opencv/pull/19384 (hasn't been released yet).

**Comment by carlocab** [comment]
`opencv` should now support this version of `tbb`.

Also looks like you might have a patch for `ceres-solver`. Not sure if it'll need backporting though.

File: Formula/tbb.rb [hunk 1]
```diff
@@ -19,50 +17,74 @@ class Tbb < Formula
...
+  # See https://github.com/oneapi-src/oneTBB/issues/343
+  patch :DATA
 
   def install
-    compiler = (ENV.compiler == :clang) ? "clang" : "gcc"
-    system "make", "tbb_build_prefix=BUILDPREFIX", "compiler=#{compiler}"
-    lib.install Dir["build/BUILDPREFIX_release/*.dylib"]
+    args = *std_cmake_args + %w[
+      -DTBB_TEST=OFF
+      -DTBB4PY_BUILD=ON
```
Comment (carlocab) [inline Formula/tbb.rb:27]: Add
```
-DCMAKE_INSTALL_RPATH=#{lib}
```
?

Comment (carlocab) [reply Formula/tbb.rb:27]: Actually, this doesn't look like the problem. I think we need to pass an `rpath` to the linker in the Python build. There's another formula that does this; if I can only remember which one 😅

Comment (carlocab) [reply Formula/tbb.rb:27]: Try this

https://github.com/Homebrew/homebrew-core/blob/8bfc48e6caa277cd8680bf5e8931e31de22d62ef/Formula/coin3d.rb#L67-L70

Comment (bayandin) [reply Formula/tbb.rb:27]: Thanks, let's give it a try

Comment (carlocab) [reply Formula/tbb.rb:27]: It worked 🎉  Still some other failures though

**Comment by bayandin** [comment]
- `brew test --retry --verbose parallelstl` — I guess we need to update it to https://github.com/oneapi-src/oneDPL/releases/tag/oneDPL-2021.2.0-release (yep, they have moved the project from github.com/intel/parallelstl)
- `brew install --verbose --build-bottle opencascade` — could be solved by https://github.com/Homebrew/homebrew-core/pull/74802, but I doubt it
- `brew install --verbose --build-bottle opencv@3` — could be solved by https://github.com/Homebrew/homebrew-core/pull/74800, but I'm not sure that they have backported https://github.com/opencv/opencv/pull/19384 to version 3.4 yet. So I guess no and we either need to wait for it, or add a patch
- `brew install --verbose --build-bottle openvdb`
- `brew install --verbose --build-bottle osrm-backend`

**Comment by carlocab** [comment]
`openvdb` and `osrm-backend` just need a little help finding `tbb`. That can be done with some extra CMake flags.

This is a bit concerning though:
```
-- Set Architecture to x64 on OS X
```
https://github.com/Homebrew/homebrew-core/pull/72682/checks?check_run_id=2295732815#step:7:2353

**Comment by carlocab** [comment]
Let's give this another go.

**Comment by carlocab** [comment]
Same six failures as before:

```
Error: 6 failed steps!
brew test --retry --verbose open-image-denoise
brew test --retry --verbose parallelstl
brew install --verbose --build-bottle opencascade
brew install --verbose --build-bottle opencv@3
brew install --verbose --build-bottle openvdb
brew install --verbose --build-bottle osrm-backend
```

**Comment by carlocab** [comment]
Same five failures (minus `open-image-denoise`). Ok, out of ideas.

**Comment by cho-m** [comment]
I guess the only options here are to:
- contribute upstream fixes (or manually backport existing fixes) for new API/header files 
- add a `tbb@2020` formula
- keep waiting until all upstreams update to newer `tbb`.

**Comment by carlocab** [comment]
`gh` keeps pulling the wrong refs so fixing it to push to this branch is too annoying. Opened #77662 instead.

**Pull request closed without merging**